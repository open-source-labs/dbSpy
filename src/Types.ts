
export type DefaultErr = {
  log: string,
  status: number,
  message: string,
};

export type Props = {
  isActive: boolean;
  setIsActive: (active: boolean) => void;
};

export interface dbCredentials{
  database_name: string|number|null;
  username: string|number|null;
  password: string|number|null;
  hostname: string|number|null;
  port: string|number|null;
  database_link: string|number|null;
  db_type: string;
}

export interface Data{
  edges: RowsOfData[];
  table: [string, RowsOfData[]];
}

export interface RefObj {
    IsDestination: boolean,
    PrimaryKeyName: string,
    PrimaryKeyTableName: string,
    ReferencesPropertyName: string,
    ReferencesTableName: string,
    constraintName: string,
  }

export interface TableColumn {
  Field?: string;
  Type?: string;
  Null?: string;
  Key?: string;
  Default?: any;
  Extra?: string;
  References?: RefObj[];
  TableName?: string;
  IsForeignKey?: boolean;
  IsPrimaryKey?: boolean;
  Value?: null;
  additional_constraints?: string | null;
  data_type?: string;
  field_name?: string;
  Name?: string;
  [key: string]: any;
  update_rule?: string;
  delete_rule?: string;
  default_type?: string;
};

export interface OracleSchema {
TABLE_NAME: string,
COLUMN_NAME: string,
DATA_TYPE: string,
DATA_DEFAULT: string | null,
CHARACTER_MAXIMUM_LENGTH: string | number | null,
IS_NULLABLE: string,
COLUMN_ID: number,
CONSTRAINT_NAME: string,
CONSTRAINT_TYPE: string,
R_TABLE_OWNER: string | null,
R_TABLE_NAME: string | null,
R_COLUMN_NAME: string | null
}

export interface TableColumns {
  [columnName: string]: TableColumn;
};

export interface TableSchema {
  [tableName: string]: TableColumns;
};

export interface ReferenceType {
  [index: number]: {
    IsDestination: boolean,
    PrimaryKeyName: string,
    PrimaryKeyTableName: string,
    ReferencesPropertyName: string,
    ReferencesTableName: string,
    constraintName: string,
  },
  length: number,
};

export type Edge = {
  id: string;
  source: string;
  sourceHandle: string;
  target: string;
  targetHandle: string;
  animated: boolean;
  label: string;
  style: { strokeWidth: number; stroke: string };
  markerEnd: {
    type: string;
    orient: string;
    width: number;
    height: number;
    color: string;
  };
};

export type DataNode = {
  id: string;
  type: 'table';
  position: { x: number; y: number };
  data: DataNodeData;
};

export interface DataNodeData {
  table: TableTuple;
  edges: Edge[];
}

export type TableTuple = 
[ TableKey: string, 
  ColumnData: { [ColumnName: string]: ColumnSchema } | RowsOfData[] | RowsOfData];

  export interface dataSourceConnection {
    type: string,
    host?: string,
    port?: string | number,
    username?:string,
    password?: string,
    database?: string,
    serviceName?: string,
    synchronize?: boolean,
    logging?: boolean,
}

// ---------------------------------------------------------------------
// ZUSTAND STORE AND FRONT-END TYPES

// ColumnData is input by user and sent to schemaStore
export type ColumnData = {
  name: string;
  type: SQLDataType;
  isNullable: boolean;
  isPrimary: boolean;
  // Using `string | null` instead of optional `?`
  // because default value can be added, which throws controlled type error
  defaultValue: string | null;
};

// ColumnSchema and Reference are used by schemaStore
export type Reference = {
  [tableName: string]: {
    PrimaryKeyName: string;
    PrimaryKeyTableName: string;
    ReferencesPropertyName: string;
    ReferencesTableName: string;
    IsDestination: boolean;
    constraintName: string;
  }
};

// export type PrimaryKeyReference = {
//   [tableName: string]: RowsOfData
// }

export type ForeignKeyData = {
  PrimaryKeyTableName: string,
  PrimaryKeyColumnName: string,
  ForeignKeyTableName: string,
  ForeignKeyColumnName: string,
  constraintName: string
}

export type InnerReference = {
  PrimaryKeyName: string;
  PrimaryKeyTableName: string;
  ReferencesPropertyName: string;
  ReferencesTableName: string;
  IsDestination: boolean;
  constraintName: string;
}

export interface ColumnSchema{
  Name: string;
  Value: string | null;
  TableName: string;
  References: InnerReference[];
  IsPrimaryKey: boolean;
  IsForeignKey: boolean;
  field_name: string;
  data_type: SQLDataType;
  additional_constraints: 'NULL' | 'NOT NULL' | 'PRIMARY' | 'UNIQUE' | '' ;
};

// these are for data tables ######################
export type RowsOfData = {
  [key: string | number]: string | number | boolean | null,
};

export type DataStore = {
  [TableName: string]: RowsOfData[];
}

export type SchemaStore = {
  [TableName: string]: {
    [ColumnName: string]: ColumnSchema;
  };
};

export type DataRowArray = Array<string | number | boolean>

// export type DataState = {
//   // DATA
//   dataStore: DataStore;
//   system: 'PostgreSQL' | 'MySQL' | 'Microsoft SQL' | 'Oracle SQL';
//   history: DataStore[];
//   historyCounter: number;

  

//   // DATA SETTERS
//   setDataStore: (dataInfo: DataStore) => void;
//   setSystem: (system: DataStore) => void;
// }

export type FlowState = {
  edges: any[];
  setEdges: (eds: any) => void;
  nodes: any[];
  setNodes: (nds: any) => void;
  onNodesChange: (changes: any) => void;
  onEdgesChange: (changes: any) => void;
  onConnect: (connection: any) => void;
};

export type ColumnDataForDataTable = {
  [key: string | number]: RowsOfData[],
  
};
//######################

export interface Table {
  [key: string]: ColumnSchema;
}

export interface SchemaObject {
  [key: string]: Table;
}

export type SQLDataType = 
  'AUTO_INCREMENT'
  | 'SERIAL'
  | 'SMALLSERIAL'
  | 'BIGSERIAL'
  | 'INT'
  | 'INT2'
  | 'INT4'
  | 'INT8'
  | 'SMALLINT'
  | 'INTEGER'
  | 'BIGINT'
  | 'DECIMAL'
  | 'NUMERIC'
  | 'REAL'
  | 'FLOAT'
  | 'FLOAT4'
  | 'FLOAT8'
  | 'DOUBLE PRECISION'
  | 'MONEY'
  | 'CHARACTER VARYING(8)'
  | 'VARCHAR(255)'
  | 'CHARACTER(8)'
  | 'CHAR(8)'
  | 'TEXT'
  | 'CITEXT'
  | 'HSTORE'
  | 'BYTEA'
  | 'BIT'
  | 'VARBIT'
  | 'BIT VARYING'
  | 'TIMETZ'
  | 'TIMESTAMPTZ'
  | 'TIMESTAMP'
  | 'TIMESTAMP WITHOUT TIME ZONE'
  | 'TIMESTAMP WITH TIME ZONE'
  | 'DATE'
  | 'TIME'
  | 'TIME WITHOUT TIME ZONE'
  | 'TIME WITH TIME ZONE'
  | 'INTERVAL'
  | 'BOOL'
  | 'BOOLEAN'
  | 'ENUM'
  | 'POINT'
  | 'LINE'
  | 'LSEG'
  | 'BOX'
  | 'PATH'
  | 'POLYGON'
  | 'CIRCLE'
  | 'CIDR'
  | 'INET'
  | 'MACADDR'
  | 'TSVECTOR'
  | 'TSQUERY'
  | 'UUID'
  | 'XML'
  | 'JSON'
  | 'JSONB'
  | 'INT4RANGE'
  | 'INT8RANGE'
  | 'NUMRANGE'
  | 'TSRANGE'
  | 'TSTZRANGE'
  | 'DATERANGE'
  | 'GEOMETRY'
  | 'GEOGRAPHY'
  | 'CUBE'
  | 'LTREE';

  export type PostgresDataTypes = 
  'bigint' 
  | 'bigserial'
  | 'bit'
  | 'bit varying'
  | 'boolean'
  | 'bool'
  | 'box' 
  | 'bytea' 
  | 'char'
  | 'character'
  | 'character varying'
  | 'cidr'
  | 'circle'
  | 'date'
  | 'decimal'
  | 'double precision'
  | 'float4'
  | 'float8'
  | 'inet'
  | 'int'
  | 'int2'
  | 'int4'
  | 'int8'
  | 'integer'
  | 'interval'
  | 'json'
  | 'jsonb'
  | 'line'
  | 'lseg'
  | 'macaddr'
  | 'macaddr8'
  | 'money'
  | 'numeric'
  | 'path'
  | 'pg_lsn'
  | 'pg_snapshot'
  | 'point'
  | 'polygon'
  | 'real'
  | 'serial'
  | 'serial2'
  | 'serial4'
  | 'smallint'
  | 'smallserial'
  | 'text'
  | 'time'
  | 'timetz'
  | 'timestamp'
  | 'timestamptz'
  | 'tsquery'
  | 'tsvector'
  | 'txid_snapshot'
  | 'uuid'
  | 'xml';
//from canvas.tsx

// export interface CanvasProps {
//   fetchedData: {
//     [key: string]: {
//       [key: string]: {
//         IsForeignKey: boolean;
//         IsPrimaryKey: boolean;
//         Name: string;
//         References: any[];
//         TableName: string;
//         Value: string | boolean | number; //originally any, check if value works
//         additional_constraints: string | null;
//         data_type: string;
//         field_name: string;
//       };
//     };
//   };
//   tableId?: string; //originally any, check if value works
//   isLoadingProps: boolean;
//   isErrorProps: boolean;
//   setFetchedData: (fetchedData: object) => void;
//   setSideBarOpened: (param: boolean) => void;
//   reference: any;
//   setSqlOpen: (sqlOpen: boolean) => void;
//   sqlOpen: boolean;
// }

// //from displayheader.tsx
// export interface DisplayHeaderProps {
//   menuPopUpOpened: boolean;
//   name: string | null;
//   picture: string | null | undefined;
//   setMenuPopUpOpened: (opened: boolean) => void;
//   setUser: (user: object /*originally any */) => void;
// }

// //from featuretab.tsx
// export interface FeatureTabProps {
//   setTablename: (e: string) => void;
//   fetchedData: {};
//   setFetchedData: (e: {}) => void;
//   setSideBarOpened: (param: boolean) => void;
//   screenshot: any;
// }

// //from menupopup.tsx
// export interface MenuPopUpProps {
//   opened: boolean;
//   setOpened: (opened: boolean) => void;
// }

// //from sidebar.tsx
// export interface SideBarProps {
//   isLoadingProps: boolean;
//   isErrorProps: boolean;
//   mutate: (data: object) => void;
//   sideBarOpened: boolean;
//   setSideBarOpened: (param: boolean) => void;
// }

// // from homeloggedinnavbar.tsx
// export interface HomeLoggedInNavbarProps {
//   user: {
//     email: string | null;
//     id: string | null;
//     name: string | null;
//     picture: string | null;
//   };
//   setUser: (user: object /*originally any */) => void;
// }
