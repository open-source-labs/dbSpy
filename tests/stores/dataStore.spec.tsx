import useDataStore, { DataObj, DataState } from '../../src/store/dataStore';
import { act, renderHook } from '@testing-library/react';
import { DataStore } from '@/Types';
import { mock } from 'node:test';

describe('Unit testing dataStore', () => {
  let result: any;
  let setDataStore: any;
  let setSystem: any;
  let deleteTableData: any;
  let addtableData: any;
  let setReferencesStore: any;

  beforeEach(() => {
    result = renderHook(useDataStore).result;
    setDataStore = result.current.setDataStore;
    setSystem = result.current.setSystem;
    deleteTableData = result.current.deleteTableData;
    addtableData = result.current.addTableData;
    setReferencesStore = result.current.setReferencesStore;
  });

  afterEach(() => {
    act(() => {
      setSystem('PostgreSQL');
      setDataStore({});
    });
  });

  describe('dataStore shape', () => {
    it('matches the snapshot', () => {
      const dataStore = renderHook(useDataStore).result.current as DataStore;
      expect(dataStore).toMatchSnapshot();
    });
  });

  describe('dataStore state changes', () => {
    it('should initialize with default state', () => {
      const { dataStore, system } = result.current;
      expect(dataStore).toEqual({});
      expect(system).toEqual('PostgreSQL');
    });
    
    it('should set dataStore state', () => {
      act(() => {
        setDataStore({ table1: {} });
      });
        
      expect(result.current.dataStore).toEqual({ table1: {} });
    });

    it('should delete table data', () => {
      act(() => {
        setDataStore({ table1: {}, table2: {} });
        deleteTableData('table1');
      });

      expect(result.current.dataStore).toEqual({ table2: {} });
    });

    
    it ('should add table data', () => {
      act(() => {
        addtableData('table1', [{ key: 'test' }]);
      });

      expect(result.current.dataStore.table1).toEqual({ table1: { key: 'test' } });
    });

    it('should handle empty newRow array', () => {
      const testTable = 'test';
      const newRow: DataObj[] = [];
      const initialDataStore: DataStore = {
        [testTable]: [{ column1: 'test', column2: 'data' }]
      };

      act(() => {
        setDataStore(initialDataStore);
      });

      act(() => {
        result.current._addRow(result.current.dataStore, testTable, newRow);
      });

      expect(result.current.dataStore[testTable]).toEqual([{ column1: 'test', column2: 'data' }]);
    });
  });

  // System should be able to be set to any of the specified SQL systems
  describe('System setting', () => {
    const systems = [ 'PostgreSQL', 'MySQL', 'Microsoft SQL', 'Oracle SQL', 'SQLite' ]

    // Reduces verbosity of writing tests for each system
    systems.forEach(system => {
      it(`should set system to ${system}`, () => {
        act(() => {
          setSystem(system);
        });
    
        expect(result.current.system).toEqual(system);
      });
    });
  });
  
  describe('References store setting', () => {
    it('should update referenceStore state', () => {
      const mockReferenceStore = {
        table1: [{ column1: 'test', column2: 'data' }]
      };

      act(() => {
        setReferencesStore(mockReferenceStore);
      });
      
      expect(result.current.referenceStore).toEqual(mockReferenceStore);
    });
  });
});