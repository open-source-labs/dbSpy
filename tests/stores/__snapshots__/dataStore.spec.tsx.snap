// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit testing dataStore dataStore shape matches the snapshot 1`] = `
{
  "_addRow": [Function],
  "addTableData": [Function],
  "dataStore": {},
  "deleteTableData": [Function],
  "referenceStore": {},
  "setDataStore": [Function],
  "setReferencesStore": [Function],
  "setSystem": [Function],
  "system": "PostgreSQL",
}
`;

exports[`Unit testing for dataStore dataStore shape matches the snapshot 1`] = `
{
  "_addRow": [Function],
  "addTableData": [Function],
  "dataStore": {},
  "deleteTableData": [Function],
  "referenceStore": {},
  "setDataStore": [Function],
  "setReferencesStore": [Function],
  "setSystem": [Function],
  "system": "PostgreSQL",
}
`;
